### PR TITLE
[3.x] Serializable & cacheable entities

### PR DESCRIPTION
**Pull request by GawainLynch**
## Problem A

Due to legacy (in most cases) entity objects in a lot of cases can't `serialize()`, e.g. for things like caches. 
as mostly for legacy reasons. they contain closures like `Application` (e.g. by `Config` being needed via proxy for taxonomy, or `EntityManager for relations`).

This is causing breakage in a number of areas, and causing me to chase :ghost: on two upcoming branches :smile: 

I am seeing broken serialised objects in session caches when the 1% random log-out that isn't caused by config settings, is **actually** serialisation breaking.

Also, as an example of where this is blocking things… 75-90% of the current dashboard loading time (assuming warm caches) is data that could be tag-cached, even in memory … but instead a ridiculous amount of logic in re-re-re-repeated on every page load. Of the example, 3/4 of that is just the menus … see where I am going with this :wink: 

## Problem B

`EntitySerializeTrait` provides the interface contract methods for entity serialisation

`JsonSerializable` :+1:, but not for `Serializable` .

Obviously adding `unserialize()` is a proverbial no-brainer.

However, currently `serialize()` returns an _array_, the interface requires the implementer to do the pre-processing, and return a serialised _string_. Changing that becomes a BC break.

## Proposed Solution

OK, let's talk … This PR makes the serialisation work, but does potentially break some corner-case use. 

There is probably something else I have missed, and am not "sold" on the chosen function name/use of the renamed functions in the collections classes

### CODE REVIEW

**Comment by bobdenotter** [comment]
Shouldn't we altogether avoid `serialize()` because of security concerns? As in: if somebody manages to pass in an object to `unserialize`, PHP will invoke its `__wakeup()` with arbitrary code.

**Comment by GawainLynch** [comment]
> Shouldn't we altogether avoid serialize

OK, so that rules out using Doctrine, Symfony & Laravel cache components

**Comment by GawainLynch** [comment]
@rossriley if you think the BC break is unacceptable, @CarsonF has suggested switching to detecting what's calling `serialize()`, and responding appropriately … I just have been too short on time to update you, or this PR :smile:

**Comment by CarsonF** [comment]
Yeah let's create a new method `toArray` or something and move all internal calls to that. Then we can detect who's calling serialize by checking if `file` is defined in the first frame of a backtrace. If there is then trigger warning and call toArray. Else serialize and return string.

**Comment by GawainLynch** [comment]
Updated as requested.

File: src/Storage/Collection/LazyCollection.php [hunk 1]
```diff
@@ -26,16 +29,50 @@ public function add($element)
...
     }
+
+    /**
+     * @internal
+     */
+    public function serialize()
+    {
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+        if (isset($trace[0]['file'])) {
+            Deprecated::warn('Calling serialize() on collections to return an array', 3.4, 'Use toArray() instead.');
```
Comment (CarsonF) [inline src/Storage/Collection/LazyCollection.php]: You could use method instead. 
```php
Deprecated::method(3.4, 'toArray');
```

File: src/Storage/Entity/Entity.php [hunk 2]
```diff
@@ -73,8 +77,36 @@ public function setId($id)
...
         return (string) $this->getId();
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toArray()
+    {
+        $data = [];
+        foreach ($this as $k => $v) {
```
Comment (GawainLynch) [inline src/Storage/Entity/Entity.php]: It's not my code, just moved.

File: src/Storage/Collection/LazyCollection.php (hunk 1 above)
Comment (GawainLynch) [inline src/Storage/Collection/LazyCollection.php]: I was aiming for more context, but fine. Done.

File: src/Storage/Collection/Taxonomy.php [hunk 3]
```diff
@@ -241,7 +245,10 @@ public function getSortorder($field, $slug)
         }
     }
 
-    public function serialize()
+    /**
+     * {@inheritdoc}
+     */
+    public function toArray()
     {
```
Comment (GawainLynch) [inline src/Storage/Collection/Taxonomy.php]: Ugh, I want to argue you down on this one … and I'm not 100% convinced of the LSP break. However, that's a moot point in my mind as your follow-up question/point makes the most sense … 

So done!

File: src/Storage/Entity/Entity.php (hunk 2 above)
Comment (CarsonF) [inline src/Storage/Entity/Entity.php]: @rossriley?

File: src/Storage/Field/Collection/RepeatingFieldCollection.php [hunk 4]
```diff
@@ -283,13 +285,33 @@ public function getEmptySet()
...
-        return $output;
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+        if (isset($trace[0]['file'])) {
+            Deprecated::method(3.4, 'toArray');
+
+            return $output;
+        }
+
+        return serialize($output);
+    }
```
Comment (CarsonF) [inline src/Storage/Field/Collection/RepeatingFieldCollection.php]: What about `$em`, `$mapping`, and `$name` properties?

File: src/Storage/Collection/Taxonomy.php [hunk 5]
```diff
@@ -204,7 +208,7 @@ public function getField($fieldName)
      */
     public function filter(Closure $p)
     {
-        $elements = $this->toArray();
+        $elements = parent::toArray();
```
Comment (CarsonF) [inline src/Storage/Collection/Taxonomy.php]: change isn't needed.

File: src/Storage/Collection/Relations.php [hunk 6]
```diff
@@ -234,13 +236,39 @@ public function offsetGet($offset)
...
+        }
+
+        $data = [];
+        foreach ($this as $element) {
+            $data[] = $element;
         }
 
-        return $output;
+        return serialize($data);
+    }
```
Comment (CarsonF) [inline src/Storage/Collection/Relations.php]: Where is the change to not call this method anymore? And all serialize() methods for that matter.

File: src/Storage/Collection/Relations.php [hunk 7]
```diff
@@ -122,7 +124,7 @@ public function update(Relations $collection)
     public function getGrouped()
     {
         $types = [];
-        $elements = $this->toArray();
+        $elements = parent::toArray();
```
Comment (CarsonF) [inline src/Storage/Collection/Relations.php]: Change is not needed.

File: src/Storage/Collection/Relations.php [hunk 8]
```diff
@@ -234,13 +236,39 @@ public function offsetGet($offset)
...
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+        if (isset($trace[0]['file'])) {
+            Deprecated::method(3.4, 'toArray');
+
+            $output = [];
+            foreach ($this as $k => $existing) {
+                $output[$existing->getToContenttype()][] = spl_object_hash($existing);
+            }
+
+            return $output;
```
Comment (CarsonF) [inline src/Storage/Collection/Relations.php]: This logic is different than what `toArray` would return. We should not suggest it or move this logic to another method someone can call.

File: src/Storage/Collection/Taxonomy.php [hunk 9]
```diff
@@ -241,14 +245,40 @@ public function getSortorder($field, $slug)
...
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+        if (isset($trace[0]['file'])) {
+            Deprecated::method(3.4, 'toArray');
+
+            return $output;
+        }
+
+        $data = [
+            'elements' => $this->toArray(),
+            'config'   => method_exists($this->config, 'getData') ? $this->config->getData() : $this->config,
```
Comment (CarsonF) [inline src/Storage/Collection/Taxonomy.php]: `ConfigurationValueProxy` should be made `Serializable` instead.

File: src/Storage/Collection/LazyCollection.php [hunk 10]
```diff
@@ -26,16 +29,50 @@ public function add($element)
...
     {
         $output = [];
         foreach ($this as $element) {
+            /** @var Content $proxy */
             $proxy = $element->getProxy();
             $output[] = $proxy->getContenttype() . '/' . $proxy->getSlug();
         }
 
         return $output;
     }
```
Comment (CarsonF) [inline src/Storage/Collection/LazyCollection.php]: This should return EntityProxies...or I guess the objects they are proxing..maybe..as that's what every other method on the class works with. 

With the way this is currently coded `get()` could return an object, an EntityProxy, or a string. Huge red flag 👎

File: src/Storage/Field/Collection/RepeatingFieldCollection.php (hunk 4 above)
Comment (GawainLynch) [inline src/Storage/Field/Collection/RepeatingFieldCollection.php]: `$em` & `$mapping` are not serializable

File: src/Storage/Collection/Taxonomy.php (hunk 5 above)
Comment (GawainLynch) [inline src/Storage/Collection/Taxonomy.php]: Yeah, forgot to pull those … I'll grab 'em in the morning

File: src/Storage/Collection/Taxonomy.php (hunk 9 above)
Comment (GawainLynch) [inline src/Storage/Collection/Taxonomy.php]: WFM

File: src/Storage/Collection/Relations.php (hunk 7 above)
Comment (GawainLynch) [inline src/Storage/Collection/Relations.php]: As above, derp and changing incoming

File: src/Storage/Field/Collection/RepeatingFieldCollection.php (hunk 4 above)
Comment (CarsonF) [inline src/Storage/Field/Collection/RepeatingFieldCollection.php]: Then neither is this class.

File: src/Storage/Collection/LazyCollection.php (hunk 10 above)
Comment (GawainLynch) [inline src/Storage/Collection/LazyCollection.php]: Yeah, good point actually … Let me dig.

File: src/Storage/Collection/Taxonomy.php (hunk 9 above)
Comment (GawainLynch) [inline src/Storage/Collection/Taxonomy.php]: Oh, so much nicer already for that one, thanks :smile:

File: src/Storage/Collection/Relations.php (hunk 6 above)
Comment (GawainLynch) [inline src/Storage/Collection/Relations.php]: That, currently, I need feedback from @rossriley on

File: src/Storage/Entity/Entity.php (hunk 2 above)
Comment (GawainLynch) [inline src/Storage/Entity/Entity.php]: (Ignore the visual references to 3.3, it is just my 3.3 worktree checked out with this branch)

entity backtrace

Comment (GawainLynch) [inline src/Storage/Entity/Entity.php]: … and believe it or not, the `getUsers()` thing was BC, no idea if what was requiring it to maintain the BC is fixed, and the class, globals, and a tonne else, is going in `master` already or soon.

File: src/Storage/Collection/LazyCollection.php (hunk 10 above)
Comment (GawainLynch) [inline src/Storage/Collection/LazyCollection.php]: OK, removed the interface and implementation methods, and changed the name of that to `getProxy()` to match `\Bolt\Storage\EntityProxy` … I'll keep digging as to what's calling what internally.

File: src/Storage/Collection/Relations.php (hunk 6 above)
Comment (GawainLynch) [inline src/Storage/Collection/Relations.php]:
```shell
$ ag "'serialize'" src/
src/Storage/Field/Type/FieldTypeBase.php
94:        $valueMethod = 'serialize' . ucfirst($this->camelize($attribute));

src/Storage/Field/Type/TemplateFieldsType.php
81:        $valueMethod = 'serialize' . ucfirst($key);

src/Storage/Migration/Export.php
113:                if (is_callable([$val, 'serialize'])) {

src/Storage/Entity/MagicAttributeTrait.php
76:        if (strncasecmp($method, 'serialize', 9) == 0) {

src/Storage/Entity/Entity.php
101:            $method = 'serialize' . $k;
106:            $method = 'serialize' . $k;

src/Storage/Entity/TemplateFields.php
44:            if (is_callable([$val, 'serialize'])) {
```

```shell
$ ag ">serialize\(" src/
src/Composer/Package/Dependency.php
62:        return $this->serialize();
72:        return $this->serialize();

src/Composer/Package/DependencyRelationship.php
30:        return $this->serialize();
40:        return $this->serialize();

src/Storage/Field/Type/TemplateFieldsType.php
88:            $value = $this->serialize($value, $metadata);

src/Storage/Field/Collection/RepeatingFieldCollection.php
295:            $output[$collection] = $vals->serialize();

src/Storage/Migration/Export.php
114:                    $val = $val->serialize();

src/Storage/Entity/TemplateFields.php
45:                $val = $val->serialize();

src/Session/SessionStorage.php
379:        $data = $this->serializer->serialize($this->data);
```

Comment (GawainLynch) [inline src/Storage/Collection/Relations.php]: So basically [`\Bolt\Storage\Entity\MagicAttributeTrait::__call()`](https://github.com/bolt/bolt/blob/release/3.4/src/Storage/Entity/MagicAttributeTrait.php#L76-L80) is catching method called, or prefixed with, "serialize" …

**Pull request closed without merging**